src/osdc/Journaler.cc

```diff
@@ -970,7 +970,6 @@ void Journaler::_issue_read(uint64_t len)
     if (flush_pos == safe_pos) {
       _flush(NULL);
     }
-    assert(flush_pos > safe_pos);
```

Comment:
flush_pos only advances in _do_flush if it's not also busy prezeroing, so in that case the assertion would fail.  It's okay though, because the waitfor_safe callback will still happen, because in _finish_prezero we check waiting_for_zero and call _do_flush again.

We issue reads at an offset that hasn't been written in the case that PurgeQueue has no backlog: in push() we call _consume().  In a busy system, this _consume will not touch Journaler because PurgeQueue::in_flight will already be at its limit.  Otherwise, _consume will call Journaler::is_readable(), which will bring us through this _issue_read path.

This could all be more efficient if PurgeQueue had a special case for when it's not busy, which would pass the item straight through to be executed right after it's been written to the journaler, without issuing a read on the journaler.